subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c

```diff
@@ -214,11 +215,205 @@ fs_mgmt_file_upload(struct mgmt_ctxt *ctxt)
+	};
+
+	type = FS_MGMT_CHECKSUM_COUNT;
+	rc = cbor_read_object(&ctxt->it, size_attr);
+
+	if (rc != 0 || type >= FS_MGMT_CHECKSUM_COUNT) {
+		return MGMT_ERR_EINVAL;
+	}
+
+	/* Calculate checksum of file */
+	rc = fs_mgmt_impl_checksum(path, (enum fs_mgmt_checksum_t)type,
+				   &checksum, &file_len, &checksum_len);
+
+	if (rc != 0) {
+		return rc;
+	}
+
+	/* Encode the response. */
+	err = 0;
+	err |= cbor_encode_text_stringz(&ctxt->encoder, "rc");
+	err |= cbor_encode_int(&ctxt->encoder, MGMT_ERR_EOK);
+	err |= cbor_encode_text_stringz(&ctxt->encoder, "type");
+	err |= cbor_encode_uint(&ctxt->encoder, type);
+	err |= cbor_encode_text_stringz(&ctxt->encoder, "len");
+	err |= cbor_encode_uint(&ctxt->encoder, file_len);
+	err |= cbor_encode_text_stringz(&ctxt->encoder, "checksum");
+	err |= cbor_encode_uint(&ctxt->encoder, checksum);
+
+	if (err != 0) {
+		return MGMT_ERR_ENOMEM;
+	}
+
+	return 0;
+#else
+	/* Feature not supported */
+	return MGMT_ERR_ENOTSUP;
+#endif
+}
+
+/**
+ * Command handler: fs hash (read)
+ */
+static int
+fs_mgmt_file_hash(struct mgmt_ctxt *ctxt)
+{
+#if CONFIG_FS_MGMT_HASH_CHUNK_SIZE > 0
+	char path[CONFIG_FS_MGMT_PATH_SIZE + 1];
+	CborError err;
+	size_t file_len;
+	size_t hash_len;
+	int rc;
+	unsigned long long type;
+	uint8_t hash[FS_MGMT_SHA256_DIGEST_SIZE];
+
+	const struct cbor_attr_t size_attr[] = {
+		{
+			.attribute = "type",
+			.type = CborAttrUnsignedIntegerType,
+			.addr.uinteger = &type,
+			.nodefault = true
+		},
+		{
+			.attribute = "name",
+			.type = CborAttrTextStringType,
+			.addr.string = path,
+			.len = sizeof(path),
+		},
+		{ 0 },
+	};
+
+	type = FS_MGMT_CHECKSUM_COUNT;
+	rc = cbor_read_object(&ctxt->it, size_attr);
+
+	if (rc != 0 || type >= FS_MGMT_HASH_COUNT) {
+		return MGMT_ERR_EINVAL;
+	}
+
+	/* Calculate hash of file */
+	rc = fs_mgmt_impl_hash(path, (enum fs_mgmt_hash_t)type, hash, &file_len, &hash_len);
+	if (rc != 0) {
+		return rc;
+	}
+
+	/* Encode the response. */
+	err = 0;
+	err |= cbor_encode_text_stringz(&ctxt->encoder, "rc");
+	err |= cbor_encode_int(&ctxt->encoder, MGMT_ERR_EOK);
+	err |= cbor_encode_text_stringz(&ctxt->encoder, "type");
+	err |= cbor_encode_uint(&ctxt->encoder, type);
+	err |= cbor_encode_text_stringz(&ctxt->encoder, "len");
+	err |= cbor_encode_uint(&ctxt->encoder, file_len);
+	err |= cbor_encode_text_stringz(&ctxt->encoder, "hash");
+	err |= cbor_encode_byte_string(&ctxt->encoder, hash, hash_len);
+
+	if (err != 0) {
+		return MGMT_ERR_ENOMEM;
+	}
+
+	return 0;
+#else
+	/* Feature not supported */
+	return MGMT_ERR_ENOTSUP;
+#endif
+}
+
 static const struct mgmt_handler fs_mgmt_handlers[] = {
 	[FS_MGMT_ID_FILE] = {
 		.mh_read = fs_mgmt_file_download,
 		.mh_write = fs_mgmt_file_upload,
 	},
+	[FS_MGMT_ID_SIZE] = {
+		.mh_read = fs_mgmt_file_size,
+		.mh_write = NULL,
+	},
+	[FS_MGMT_ID_CHECKSUM] = {
+		.mh_read = fs_mgmt_file_checksum,
+		.mh_write = NULL,
+	},
+	[FS_MGMT_ID_HASH] = {
+		.mh_read = fs_mgmt_file_hash,
+		.mh_write = NULL,
+	},
```

Comment:
When they are not supported, do not add them here: the mcumgr will respond with error 8 when unable to find handler, and adding these makes it just do sequence:
1) search handler
2) call handler
3) handler returns with error
4) reset SMP buffer and form "rc" response with error returned from handler

Instead of directly going:
1) search handler;
1) send "rc" response with error 8
Additionally this adds code, when commands are not supported, that duplicates already existing behavior of SMP request processor.